dot/parachain/spam_slots.go

```diff
@@ -0,0 +1,202 @@
+	// We call this once a dispute becomes obsolete or got confirmed and thus votes for it should no longer be treated
+	// as potential spam.
+	Clear(session parachain.SessionIndex, candidate common.Hash)
+	// PruneOld prune all spam slots for sessions older than the given index.
+	PruneOld(oldestIndex parachain.SessionIndex)
+}
+
+type slotKey struct {
+	session   parachain.SessionIndex
+	validator parachain.ValidatorIndex
+}
+
+type unconfirmedKey struct {
+	session   parachain.SessionIndex
+	candidate common.Hash
+}
+
+// SpamSlotsHandler is an implementation of SpamSlots
+type SpamSlotsHandler struct {
+	slots     map[slotKey]uint32
+	slotsLock sync.RWMutex
+
+	unconfirmed     map[unconfirmedKey]*treeset.Set
+	unconfirmedLock sync.RWMutex
+
+	maxSpamVotes uint32
+}
+
+// byValidatorIndex is a comparator for ValidatorIndex
+func byValidatorIndex(a, b interface{}) int {
+	return int(a.(parachain.ValidatorIndex) - b.(parachain.ValidatorIndex))
+}
+
+func newSlotsKey(session parachain.SessionIndex, validator parachain.ValidatorIndex) slotKey {
+	return slotKey{
+		session:   session,
+		validator: validator,
+	}
+}
+
+func newUnconfirmedDisputesKey(session parachain.SessionIndex, candidate common.Hash) unconfirmedKey {
+	return unconfirmedKey{
+		session:   session,
+		candidate: candidate,
+	}
+}
+
+func (s *SpamSlotsHandler) getSpamCount(session parachain.SessionIndex, validator parachain.ValidatorIndex) (uint32, bool) {
+	s.slotsLock.RLock()
+	defer s.slotsLock.RUnlock()
+	spamCount, ok := s.slots[newSlotsKey(session, validator)]
+	return spamCount, ok
+}
+
+func (s *SpamSlotsHandler) getValidators(session parachain.SessionIndex, candidate common.Hash) (*treeset.Set, bool) {
+	s.unconfirmedLock.RLock()
+	defer s.unconfirmedLock.RUnlock()
+	validators, ok := s.unconfirmed[newUnconfirmedDisputesKey(session, candidate)]
+	return validators, ok
+}
+
+func (s *SpamSlotsHandler) clearValidators(session parachain.SessionIndex, candidate common.Hash) {
+	s.unconfirmedLock.Lock()
+	defer s.unconfirmedLock.Unlock()
+	delete(s.unconfirmed, newUnconfirmedDisputesKey(session, candidate))
+}
+
+func (s *SpamSlotsHandler) clearSlots(session parachain.SessionIndex, validator parachain.ValidatorIndex) {
+	s.slotsLock.Lock()
+	defer s.slotsLock.Unlock()
+	delete(s.slots, newSlotsKey(session, validator))
+}
+
+func (s *SpamSlotsHandler) incrementSpamCount(session parachain.SessionIndex, validator parachain.ValidatorIndex) {
+	s.slotsLock.Lock()
+	defer s.slotsLock.Unlock()
+	s.slots[newSlotsKey(session, validator)]++
+}
+
+func (s *SpamSlotsHandler) AddUnconfirmed(session parachain.SessionIndex,
+	candidate common.Hash,
+	validator parachain.ValidatorIndex) bool {
+	if spamCount, _ := s.getSpamCount(session, validator); spamCount >= s.maxSpamVotes {
+		return false
+	}
+
+	s.unconfirmedLock.Lock()
+	defer s.unconfirmedLock.Unlock()
+
+	validators, ok := s.unconfirmed[newUnconfirmedDisputesKey(session, candidate)]
```

Comment:
I see that you're calling the method twice, you could use a variable and assign the return to it and use the variable in both places
```suggestion
        unconfirmerdDisputesKey := newUnconfirmedDisputesKey(session, candidate)
	validators, ok := s.unconfirmed[unconfirmerdDisputesKey]
	...
	s.unconfirmed[unconfirmerdDisputesKey] = validators
```

Reply:
Well, that's not good